container: "docker://condaforge/mambaforge:4.10.1-0"
import pandas as pd
configfile: "config/config.yaml"

df = (
    pd.read_csv(config["samples"], sep="\t", dtype={"sample_name": str, "unit_name": str})
    .set_index(["sample_name", "unit_name"], drop=False)
    .sort_index()
)

SAMPLES=df.sample_name.unique()
samples_trim=df.sample_name
unit_trim=df.unit_name

rule all:
    input:
        "results/qc/samples_qc.html",
        "results/qc/reads_qc.html",
        "results/qc/bam_qc.html",
        expand("results/counts/featureCounts/{sample}_geneCounts_gencode.tsv", sample=SAMPLES),
        expand("results/counts/salmon/{sample}_geneCounts.tsv", sample=SAMPLES),
        expand("results/paQuant/{sample}_paQuant.tsv.gz", sample=SAMPLES),
        expand("results/fusion/arriba/{sample}.tsv", sample=SAMPLES),
        expand("results/variantCalling/vcf_filtered/{sample}.vcf.gz", sample=SAMPLES),
        expand("results/fusion/STAR_fusion/{sample}/star-fusion.fusion_predictions.tsv", sample=SAMPLES),
        "resources/GRCh37_gencode_v19_CTAT_lib_Mar012021.plug-n-play/ctat_genome_lib_build_dir/ref_genome.fa"


if config["library_type"] == "PE":
    include:"rules/PE/inputFunctions.smk"
    include:"rules/downloads.smk"
    include:"rules/PE/RNASeq.smk"
    include:"rules/PE/variantCalling.smk"
    include:"rules/PE/QC.smk"
else:
    include:"rules/SE/inputFunctions.smk"
    include:"rules/downloads.smk"
    include:"rules/SE/RNASeq.smk"
    include:"rules/SE/variantCalling.smk"
    include:"rules/SE/QC.smk"
